### DISCUSSION TITLE
Does removing a SubShape change the CompoundShape Shapes Array?

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#174] [Q&A] [answered]
I'm working on a version of dynamic shapes and trying to figure out the best way to track the identity of a subshape so I can modify it at runtime.

The [ModifyShape](https://jrouwe.github.io/JoltPhysics/class_mutable_compound_shape.html#a3992cb7a984307a08139812e44da14ae) method takes the index of the shape.

What I'm worried about is the `RemoveShape` method...
Lets say I have shapes like `MyCompoundShape = [(A:0), B:1), (C:2), (D:3)]`

If I do `MyCompounShape.RemoveShape(1)`

In Javascript the shapes array would look like: `[(A:0), (C:1), (D:2),]`

Which means shapes C & D have new indexes. However I'm not sure if the C++ Array would work this way as its an `Array` and not a `Vector` so it might maintain the indexes. 

I thought about using [GetSubShapeIDFromIndex](https://jrouwe.github.io/JoltPhysics/class_compound_shape.html#a66ae55d5b737c492d73286dff4615215) and storing the subshape but it's a little confusing how this works and what params I'd pass around.

My other solution was to just set a hash in `UserData` and track that, then make a getter, but that's a lot of overhead if it isn't needed. 

If Indexes don't change I'll go with that, if they do I'll probably buckle down on the SubShapeID

**Comment by jrouwe** [comment, marked as answer]
When you call `RemoveShape(1)` the indexes shift, so you'll end up with `A:0, C:1, D:2`.

Note that the SubShapeID basically encodes a sub shape index, so if you get a SubShapeID and then remove a sub shape, it will no longer point to the same sub shape.

The `UserData` of a sub shape was meant to keep track of sub shapes while adding/removing and is the only thing that can be used reliably (although from your post I gather you want the indices to shift, so maybe you don't need to track anything extra).

**Reply by DennisSmolek** [reply]
In my application I'm not too concerned with the index on it's own but as it was a factor in getting the subShapeId.
But if it changes then I'll not really need it at all. Instead I'll create a hash and set the userData then make a getter of my own.

Thanks as always for the quick reply!
